src/tesla_client.cpp

```diff
@@ -0,0 +1,745 @@
+  req->onResponse([&](MongooseHttpClientResponse *response) {
+      DEBUG.println("resp");
+      printResponse(response);
+
+      // bad credentials - prevent locking of account
+      // by clearing the user/pass so we don't retry w/ them
+      if (response->respCode() == 401) {
+	_teslaUser = "";
+	_teslaPass = "";
+	DEBUG.println("bad user/pass");
+      }
+      else if (response->respCode() == 200) { 
+	const char *cjson = (const char *)response->body();
+	if (cjson) {
+	  String json = cjson;
+	  const size_t capacity = JSON_OBJECT_SIZE(5) + 220;
+	  DynamicJsonDocument doc(capacity);
+	  deserializeJson(doc, json);
+	  const char *token = (const char *)doc["access_token"];;
+	  if (token) {
+	    _accessToken = "Bearer ";
+	    _accessToken += token;
+	    DEBUG.print("token: ");DEBUG.println(_accessToken);
+	    _lastRequestStart = 0; // allow requestVehicles to happen immediately
+	  }
+	  else {
+	    _accessToken = "";
+	    DEBUG.println("token error");
+	  }
+	  
+	  //const char* token_type = doc["token_type"];
+	  //long expires_in = doc["expires_in"];
+	  //const char* refresh_token = doc["refresh_token"];
+	  //long created_at = doc["created_at"]; // 1583079526
+	}
+      }
+      _activeRequest = TAR_NONE;
+    });
+  _client.send(req);
+}
+
+
+void TeslaClient::requestVehicles()
+{
+  DEBUG.println("requestVehicles()");
+
+  _activeRequest = TAR_VEHICLES;
+  _lastRequestStart = millis();
+
+  String uri = TESLA_BASE_URI;
+  uri += "/api/1/vehicles/";
+  _accessToken = "Bearer f5b1e0239feb331db1fa5eb4969fc9c53a292124d797a740485dc375fc3dab50";
+
+  MongooseHttpClientRequest *req = _client.beginRequest(uri.c_str());
+  req->setMethod(HTTP_GET);
+  req->addHeader("Connection","keep-alive");
+  req->addHeader("User-Agent",_userAgent);
+  req->addHeader("Accept","*/*");
+  req->addHeader("Authorization",_accessToken.c_str());
+  req->onResponse([&](MongooseHttpClientResponse *response) {
+      DEBUG.println("resp");
+      printResponse(response);
+      
+      if (response->respCode() == 401) {
+	  _accessToken = "";
+	  _activeRequest = TAR_NONE;
+	  return;
+      }
+      else if (response->respCode() == 200) {
+	const char *json = (const char *)response->body();
+	
+	if (strstr(json,_badAuthStr)) {
+	  _accessToken = "";
+	  _activeRequest = TAR_NONE;
+	  return;
+	}
+	
+	
+	char *sc = strstr(json,"\"count\":");
+	if (sc) {
+	  _cleanVehicles();
+	  sc += 8;
+	  sscanf(sc,"%d",&_vehicleCnt);
+	  DEBUG.print("vcnt: ");DEBUG.println(_vehicleCnt);
+	  
+	  _id = new String[_vehicleCnt];
+	  _vin = new String[_vehicleCnt];
+	  _displayName = new String[_vehicleCnt];
+	  
+	  // ArduinoJson has a bug, and parses the id as a float.
+	  // use this ugly workaround.
+	  const char *sj = json;
+	  for (int v=0;v < _vehicleCnt;v++) {
+	    const char *sid = strstr(sj,"\"id\":");
+	    if (sid) {
+	      sid += 5;
+	      _id[v] = "";
+	      while (*sid != ',') _id[v] += *(sid++);
+	      sj = sid;
+	    }
+	  }
```

Comment:
Yes, I used the latest github code, and it was still broken. Wasted a lot of time on this.. ArduinoJson was broken internally.. the casts didn't change anything. I don't know if they've fixed it since

Reply:
there's a line further down:

	    // doesn't work.. returns converted float _id[i] = responsei["id"].as<String>();

When I tried int, it just took the float and converted that, which truncated it. Internally, it was storing it as a float.

You can try w/ the latest to see if they've fixed it now. Sorry, I don't want to waste any more time on this